CLEWS-24615: API client errors handling

waracci opened the pull request:
Description
- These are the errors that are handled by the added exception **400, 401, 404, 500**
- The current implementation returns a response as shown below 👇 

> _API call failed with the error message:_
> Giving up on https://api.gro-intelligence.com/sample_unit/ after 1 tries:

- Seeing as it is difficult to parse the message ☝️ the error message will include the following info in the response body

1. Status Code
2. Retry Count
3. Response
4. Message
5. Url

waracci committed:
CLEWS-24615: API client errors handling

- errors handled by exception 400, 401, 404, 500
- any of above errors throw an exception that contains descriptive info

itsjohnward commented on api/client/lib.py:
```diff
@@ -232,7 +244,10 @@ def get_data(url, headers, params=None, logger=None):
+                return error
         else:
             logger.error('{}'.format(data), extra=log_record)
     raise Exception('Giving up on {} after {} tries: {}.'.format(
```
Let's also raise an APIError here. This one's if you're timing out repeatedly and it gives up, so `data` in this case would likely be a 503 or 504 error.

waracci replied:
Done!

waracci committed:
CLEWS-24615: raise an exception to catch continous timouts

- in addition to catching 4** type errors
- this should catch other 5** type errors

itsjohnward committed:
Break retry loop for error-level status codes

itsjohnward committed:
Parse error message and add params to APIError

itsjohnward commented:
Hey @waracci FYI I just added some message parsing on the error responses to something I think is a bit more useful:

Test script:

```py
try:
    client.get_data_points(region_id=19)
except Exception as e:
    print('e {}'.format(e))
    print('response {}'.format(e.response))
    print('retry_count {}'.format(e.retry_count))
    print('url {}'.format(e.url))
    print('status_code {}'.format(e.status_code))
    print('message {}'.format(e.message))
```

Output changes from:

```py
response <Response [400]>
retry_count 1
url https://api.gro-intelligence.com/v2/data
status_code 400
message Giving up on https://api.gro-intelligence.com/v2/data after 1 tries: <Response [400]>.
```

to:

```py
response <Response [400]>
retry_count 1
url https://api.gro-intelligence.com/v2/data
status_code 400
message Bad Request: child "metricId" fails because ["metricId" is required]
```

Also I added `params` to the error object. Use case would be if you're looping through many different requests to the same endpoint and catching the errors as you go, you can access the `params` property to retroactively see which inputs resulted in the error.
